client/my-sites/controller.js

```diff
@@ -109,6 +111,43 @@ function renderNoVisibleSites( context ) {
 	);
 }
 
+function isPathAllowedForDomainOnlySite( pathname ) {
+	const urlPrefixesWhiteListForDomainOnlySite = [
+		'/domains/manage/',
+		'/checkout/',
+	];
+
+	return urlPrefixesWhiteListForDomainOnlySite.some( path => startsWith( pathname, path ) );
+}
+
+function onSelectedSiteAvailable( context ) {
+	const selectedSite = sites.getSelectedSite();
+	siteStatsStickyTabActions.saveFilterAndSlug( false, selectedSite.slug );
+	context.store.dispatch( receiveSite( selectedSite ) );
+	context.store.dispatch( setSelectedSiteId( selectedSite.ID ) );
+
+	const state = context.store.getState();
+
+	if ( isDomainOnlySite( state, selectedSite.ID )
+		&& ! isPathAllowedForDomainOnlySite( context.pathname ) ) {
+		page.redirect( '/domains/manage/' + selectedSite.slug );
```

Comment:
We have path building functions for these routes here: https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/upgrades/paths.js